### DISCUSSION TITLE
AspNet missing power for real enterprise Web app development.

### DISCUSSION THREAD

**Discussion by ghost** [dotnet/aspnetcore#27324] [General] [answered]
Hello, i am new developer with asp.net, but many years deveopment experience with Java, PHP.

When i tried to understand asp.net , and try to design and develope a real business web app. It's hard works.

For example, If we development a Online Store system.

We need build the (Admin, Sales, Customer, Front) module for each targed user.

The simple MVC of asp.net not well support for this.

There should be ( [Module, Controller, Action] +  Data Model Layer  + RPC + Tools) design

Module: each big part should be orgnization under module.

Controller : real business, such as [payment, GiftCode,  account, product list, product detail]

Action:  make a billing,  find the post code, chat with the sales. send forget password.

Data Model Layer; Accees the MySQL, Postgres, Redis, 

RPC: call other internal or third party  service.

Tools: send email, send sms notify, price calculator, event bus

Asp.net missing these part.

User may found it is hard to build a success web app within  Asp.net,  Too many things to do with it.

But the Ruby On Rails, and Spring boot / Spring cloud provide much more power to deal with these.

See spring boot family

![image](https://user-images.githubusercontent.com/66301650/97550861-2b670180-1a0d-11eb-9146-75fc3dc1418c.png)

May tools and modules can help us.

What's next to go , Asp.net.

**Comment by damianh** [comment]
> Module: each big part should be orgnization under module.

In my very honest opinion you are absolutely correct in this respect. I have two blog posts on this subject which should not have been needed to be written.

- https://dhickey.ie/2018/06/09/aspnet-core-nested-apps/
- https://dhickey.ie/2020/10/20/aspnet-core-3-nested-apps/

It's unfortunate that we lost the ability to be modular when the team moved away from OWIN.

**Reply by davidfowl** [reply]
Actually, OrchardCore supports this pretty nicely. The only downside being that it's named orchard (which has some stigma 😄).

@netroby Maybe take a look at https://www.orchardcore.net/.

**Reply by sebastienros** [reply]
I prefer @davidfowl 's answer better than the one @javiercn selected ;)

**Reply by hikalkan** [reply]
Also look at the https://abp.io/ which tries to fill the gap between the enterprise software development requirements and the plain asp.net core.
GitHub: https://github.com/abpframework/abp

**Comment by ghost** [comment]
@damianh  another problem may be : the core team member have no experience to write a huge web app, (like Amazon, NewEgg)

So they can not figure out what need to do if we want to build such a project.

**Comment by javiercn** [comment, marked as answer]
@netroby thanks for contacting us.

We provide some guidance on how to structure large applications that you can review [here](https://docs.microsoft.com/en-us/dotnet/architecture/cloud-native/) as well as [here](https://docs.microsoft.com/en-us/dotnet/architecture/microservices/).

For a general entry point you can check [here](https://dotnet.microsoft.com/learn/aspnet/microservices-architecture)

With regards to "modules" it is a very overloaded concept, but one way to map it is 1 module 1 assembly, MVC will pick modules from class libraries without issues.

WRT your comments on controllers and actions I agree with you, but that it's up to the app developer. 

The guides referenced above should provide more detail and come with a reference sample project that details it. See it [here](https://github.com/dotnet-architecture/eShopOnContainers)

For RPC we provide support for building REST APIs and gRPC services and we automatically generate clients for them. You can expose REST APIs using Open API to integrate with other third-parties. See [here for gRPC](https://docs.microsoft.com/en-us/aspnet/core/grpc/?view=aspnetcore-3.1) and [here for REST APIs with Open API](https://docs.microsoft.com/en-us/aspnet/core/tutorials/web-api-help-pages-using-swagger?view=aspnetcore-3.1)

For Data access there are multiple solutions out there with support for a wide range of providers, including the most common ones. 

We build and ship an ORM (Entity Framework) and there are other community based options available like NHibernate or Dapper amongst others. See [here](https://docs.microsoft.com/en-us/ef/core/) for EF docs.

With regards to "tools" there are many open source projects you can leverage like MailKit for sending email or mass transit for service bus or commercial offerings from Microsoft like Azure Communications Service or Azure Service Bus as well as commercial offerings from third parties.

I hope this helps bringing some clarity to some of the options that you have available in the .NET landscape and help you with some of your questions.

**Comment by ghost** [comment]
[Architecting Cloud Native .NET Applications for Azure]  If you means this guide, I will not mark it as answered.

Title may told us: For Azure.

If without the Azure, how could we do?

The other link i had already checked, thanks for put them all in one answer.

Asp.net will got better and better, we should focus on developer experience.

**Reply by javiercn** [reply]
While there is a section about Azure on the first guide, the concepts and patterns are cloud agnostic

**Comment by ghost** [comment]
I can see each article contains words [Cloud native, server less]

The real world web apps may not ready for this advance architecture.

Should it be more stable for current enterprise development?

Yes , you can clould ready. but still gave us time to convert or stay with behind.


No, company not ready for cloud native.

Your advance architecture too new to adopt for us.

**Reply by javiercn** [reply]
If you are looking for more traditional guidance you can check this [link](https://docs.microsoft.com/en-us/dotnet/architecture/modern-web-apps-azure/). The content here there should help guide you to build a more "traditional" application. You can skip the sections related to Azure if you are planning to deploy on-premise.

**Reply by ghost** [reply]
![image](https://user-images.githubusercontent.com/66301650/97650615-0f0ea780-1a95-11eb-87aa-89c0ab2b7a1b.png)

This architect looks well. It shall be the default guide to write a webapp using asp.net

https://docs.microsoft.com/en-us/dotnet/architecture/modern-web-apps-azure/develop-asp-net-core-mvc-apps

**Comment by ghost** [comment]
I really understand, there be huge walls between the .net core development team and (us , the real world programmer, company)

**Reply by javiercn** [reply]
If you have any concrete suggestion you want to make that helps bridge any gap you think it exists we will be happy to take it into consideration. We are always looking for ways to improve the platform.

**Reply by ghost** [reply]
Hello, do you see the next floor. The abp.io seems a good solution

**Comment by iancooper** [comment]
Hi @netroby I am the owner of Brighter, an OSS project that supports a range of middleware transports for messaging. Can’t fix all your asks, but we can help with that one. Hopefully you will find out approach easy if you have learned ASP.NET

https://github.com/BrighterCommand/Brighter

and there are others in the same space as us.

**Reply by ghost** [reply]
Looks good for me , It is very close to our need.

Thanks for told me. 

Any other solutions similar this ?

**Reply by iancooper** [reply]
@netroby Our main OSS .NET rivals would be Mass Transit: https://github.com/MassTransit/MassTransit and Rebus https://github.com/rebus-org/Rebus 

We have slightly different opinions on some things, and so you can really pick the one that works. (Apologies to other folks who I have probably missed).

Another path is Dapr, which is an MS offering that works via the sidecar model. It's not native .NET, but is perhaps useful in heterogeneous environments where you want to elide platform differences: https://github.com/dapr/

Commercial products like NServiceBus. are very good, it obviously depends on you need: https://particular.net/nservicebus
